content/learn/book/ecs/_index.md

```diff
@@ -0,0 +1,140 @@
+In many cases, we want to use *some* default fields, but not all of them.
+As Rust does not have default arguments, we instead use [struct update syntax](https://doc.rust-lang.org/book/ch05-01-defining-structs.html#creating-instances-from-other-instances-with-struct-update-syntax) in combination with the [`default()`](https://doc.rust-lang.org/std/default/trait.Default.html) method from the `Default` trait to allow us to quickly set some values while leaving others unchanged.
+
+```rust
+// We want to quickly generate new enemies from this component bundle
+// By using some default fields
+#[derive(Bundle, Default)]
+struct EnemyBundle {
+	enemy: Enemy,
+	life: Life,
+	attack: Attack,
+	defense: Defense,
+}
+
+// Dataless unit structs automatically have a default value
+// Since only one possible value exists
+struct Enemy;
+
+struct Life(u32);
+
+// Manually implementing Default for one of our component types
+// allowing us to control the starting value
+impl Default for Life {
+	fn default() -> Self {
+		Life(100)
+	}
+}
+
+// These components will have a default value of 0, as u32::default() == 0
+#[derive(Default)]
+struct Defense(u32);
+
+#[derive(Default)]
+struct Attack(u32);
+
+let my_enemy = EnemyBundle {
+	attack: 10,
+	defense: 5,
+	// All other fields are initialized from their default value
+	// setting enemy: Enemy and life: Life(100)
+	..Default::default()
+}
+```
+
+### Type aliases don't play nice
+
+Bey's ECS uses Rust's type system to dispatch data to our systems as requested.
+This is very convenient for safety and ergonomic reasons, but means that only one resource of each Rust type can exist at once, and only one component of each type can be stored on each entity.
+
+"One of each Rust type" has somewhat surprising consequences for new Rust users though: [type aliases](https://doc.rust-lang.org/reference/items/type-aliases.html) will not result in unique types, and should generally not be used to define resource or component types.
+
+```rust
+use std::any::type_name;
+
+// Our first component type
+struct Life(u32);
+
+// An attempt at a second component type, using a type alias
+type Health = Life;
+
+// Unfortunately, these two types share the same name :(
+// Attempting to insert both Life and Health would result in overwritten values,
+// and any query for one would return either
+assert_eq!(type_name(Life(42)), type_name(Health(42));
+```
+
+Instead, you have to define entirely new types for each component or resource you wish to use,
+commonly using the [`newtype` pattern](https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#using-the-newtype-pattern-to-implement-external-traits-on-external-types) to wrap a single value of some common or external type.
+
+```rust
+use std::any::type_name;
+
+struct Life(u32);
+struct Health(u32);
+
+// Our components have different types now!
+assert!(type_name(Life(42)) != type_name(Health(42));
+```
```

Comment:
The use of type ids instead of type names feels more appropriate to me, since that is what really distinguished two different types.

Reply:
Absolutely. Changing.